clang/test/Parser/cxx-template-argument.cpp

```diff
@@ -141,3 +141,15 @@ namespace r360308_regression {
     return a == b;
   }
 }
+
+namespace GH95598 {
+template<typename _Tp, bool _IsPtr = __is_pointer(_Tp)>
+struct __is_pointer {};
+// expected-warning@-1 {{keyword '__is_pointer' will be made available as an identifier for the remainder of the translation unit}}
```

Comment:
This diagnostic is very confusing because it implies the name is only available as an identifier when in reality it's available as *either* an identifier or a keyword. This seems to be the original functionality though: https://godbolt.org/z/WzTs4T31j (based on the changes in https://github.com/llvm/llvm-project/commit/47642d2b7ea175a6c0d80bf14354246ca20af342 which added this diagnostic).

So I think your changes are aligned with the original intent of the diagnostic, but I've CCed @zygoloid and @rnk just in case others from this era recall differently.

Reply:
I think we've incrementally ended up in a bad place. Here's my memory of what happened:

Originally, the idea was: we provide builtins such as `__is_pointer` for use by the standard library. But for a bunch of those, libstdc++ provides a type with the same name that works it out for itself. OK, we can work around that: if we see a declaration with that name, then the standard library isn't using our builtin, so we treat it as a non-keyword so that libstdc++ doesn't break.

But then it turns out that boost uses these things as keywords directly, rather than using the standard library functionality. To keep both that and libstdc++ working at the same time, we treat the keyword followed by an open paren as having the builtin meaning even when the keyword has been reverted to being a non-keyword.

And now we're noticing that we also need special behavior in other parts of the parser to treat the keyword followed by an open paren as a special case.

I don't think we should be extending this hack further. I've suggested on the libstdc++ bug report that this be fixed in libstdc++ instead, by stopping using our keywords as their type names. Maybe one day we can then remove it entirely.

Reply:
One thing we could consider doing would be removing the entire "revertible keyword" mechanism, and instead, in the lexer, classifying these tokens as keywords only if the next pp-token is an `lparen`. That would probably simplify the handling of these keywords everywhere, and would happen to give the outcome that libstdc++ is expecting in this case too. (It's a breaking change for some weird corner cases, such as producing the keyword via macro expansion and things like `bool __is_pointer(true);`, but hopefully those things don't happen in practice!)

Reply:
@zygoloid https://github.com/itanium-cxx-abi/cxx-abi/blob/07427ac5b6ff0799c47e8c754609cf8220c05cc0/gcc-tinfo2.cc#L373 :(

Reply:
@zygoloid I prototyped something https://github.com/llvm/llvm-project/pull/96097

One of the test has a default constructor declaration, ie `is_pod()`, which the lexer approach would struggle to cope with. Let me know what you think

The alternative is to keep status quo or go with the current PR and change the warning to a depreciation warning.
My concern is that until we support older libstdc++ (which will be the case for a long time), we won't be able to rip the hack out

Reply:
There are now patches posted for libstdc++ to avoid using our type trait keywords as identifiers.